src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/PasteOrdering.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands.Ordering
+{
+    /// <summary>
+    /// This does proper file ordering for pasting or dropping items into a folder or project.
+    /// </summary>
+    [Export(typeof(IPasteDataObjectProcessor))]
+    [Export(typeof(IPasteHandler))]
+    [AppliesTo(ProjectCapabilities.SortByDisplayOrder)]
+    [Order(OrderPrecedence)]
+    internal class PasteOrdering : IPasteHandler, IPasteDataObjectProcessor
+    {
+        public const int OrderPrecedence = 10000;
+
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IProjectAccessor _accessor;
+
+        private IProjectTree _dropTarget;
+
+        private IPasteHandler _pasteHandler;
+        private IPasteDataObjectProcessor _pasteProcessor;
+
+        [ImportingConstructor]
+        public PasteOrdering(UnconfiguredProject unconfiguredProject, IProjectAccessor accessor)
+        {
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(accessor, nameof(accessor));
+
+            _configuredProject = unconfiguredProject.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+            _accessor = accessor;
+
+            PasteHandlers = new OrderPrecedenceImportCollection<IPasteHandler>(projectCapabilityCheckProvider: unconfiguredProject);
+            PasteProcessors = new OrderPrecedenceImportCollection<IPasteDataObjectProcessor>(projectCapabilityCheckProvider: unconfiguredProject);
+        }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IPasteHandler> PasteHandlers { get; set; }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IPasteDataObjectProcessor> PasteProcessors { get; set; }
+
+        private IPasteHandler PasteHandler
+        {
+            get
+            {
+                if (_pasteHandler == null)
```

Comment:
PasteHandlers can change over the lifetime of a project, you can't snapshot one.

Reply:
Let me expand on this; as the capabilities of a project changes - so does the the results the OrderPrecendenceImportCollection returns.

Reply:
Ok, I won't make this lazy then.